stdlib/path.gr

```diff
@@ -0,0 +1,571 @@
+ * @returns An `Ok` variant containing the relative path if successfully parsed or `Err` otherwise
+ * 
+ * @example // relativeTo(/usr, /usr/bin) => ./bin
+ * @example // relativeTo(/home/me, /home/me) => .
+ * @example // relativeTo(/usr/bin, /etc) => ../../etc
+ * @example // relativeTo(.., ../../thing) => ../thing
+ * @example // relativeTo(/usr/bin, C:/Users) => Err(DifferentRoots)
+ * @example // relativeTo(../here, ./there) => Err(ImpossibleRelativization)
+ * 
+ * @since v0.5.5
+ */
+export let relativeTo = (source: Path<a, b>, dest: Path<a, c>) => {
+  let result = match ((source, dest)) {
+    ((Abs(r1), _, s1), (Abs(r2), _, s2)) =>
+      if (r1 != r2) Err(DifferentRoots)
+      else relativizeDepth((0, List.reverse(s1)), (0, List.reverse(s2))),
+    ((Rel(up1), _, s1), (Rel(up2), _, s2)) =>
+      relativizeDepth((up1, List.reverse(s1)), (up2, List.reverse(s2))),
+    _ => fail "",
+  }
+
+  let (_, fileType, _) = dest
+  match (result) {
+    Ok((depth, segs)) => Ok((Rel(depth), fileType, List.reverse(segs))),
+    Err(err) => Err(err),
+  }: Result<Path<Relative, c>, RelativizationError>
+}
+
+let rec segsDescendantOf = (baseSegs, pathSegs) =>
+  match ((baseSegs, pathSegs)) {
+    ([], _) => true,
+    (_, []) => false,
+    ([f1, ...r1], [f2, ...r2]) => f1 == f2 && segsDescendantOf(r1, r2),
+  }
+
+/**
+ * Determines if a path is a descendant of another path.
+ * 
+ * @param base: The path to consider as the possible ancestor path
+ * @param path: The path to consider as the possible descendant path
+ * @returns `true` if the second path is a descendant of the first path and `false` otherwise
+ * 
+ * @example // isDescendantOf(/usr, /usr/bin/bash) == true
+ * @example // isDescendantOf(/usr, /etc) == false
+ * @example // isDescendantOf(../dir1, ./dir2) == false
+ * 
+ * @since v0.5.5
+ */
+export let isDescendantOf = (base: Path<a, Directory>, path: Path<a, c>) => {
+  let (b1, _, s1) = base
+  let (b2, _, s2) = path
+  b1 == b2 && segsDescendantOf(List.reverse(s1), List.reverse(s2))
+}
+
+/**
+ * Returns the path corresponding to the parent directory of the given path.
+ * 
+ * @param path: The path to get the parent of
+ * @returns A path corresponding to the parent directory of the given path
+ * 
+ * @since v0.5.5
+ */
+export let parent = (path: Path<a, b>) =>
+  match (path) {
+    (base, _, [_, ...rest]) => (base, Dir, rest),
+    (Rel(upDirs), _, []) => (Rel(upDirs + 1), Dir, []),
+    (Abs(_) as base, _, []) => (base, Dir, []),
+  }: Path<a, Directory>
+
+/**
+ * Returns the base name (named final segment) of a path.
+ * 
+ * @param path: The path to get the name of
+ * @returns A `Some` variant containing the name of the path, or `None` if the path does not have one
+ * 
+ * @example Path.name(Option.unwrap(Path.relativeFile("./dir/file.txt"))) // Some("file.txt")
+ * @example Path.name(Option.unwrap(Path.relativeDir(".."))) // None
+ * 
+ * @since v0.5.5
+ */
+export let name = (path: Path<a, b>) =>
```

Comment:
```suggestion
export let basename = (path: Path<a, b>) =>
```
Oscar said he liked `basename` the most.

Also, wondering about this API. Should it take a `PathWithType` instead? That way you could do something like `Path.basename(Path.fromString())` instead of needing to unwrap the option.

Reply:
The assumption I was going under was that you'd usually know what type of path you'd be working with and that `PathWithType` would only be used in extenuating circumstances; therefore all of the significant parts of this API were made with the assumption that "bare" `Path`s would be passed in.

In any case, most of the other functions in this library use the bare Paths over `PathWithType` so if this change is made I feel that it would be inconsistent to only have it for this function and not others like `parent`, `toString` etc.

Reply:
Yeah, I'm struggling with the duality of having a nice API vs type safety. I really like what you've done with `absoluteFile`/`relativeFile` and `absoluteDir`/`relativeDir` but it trades safety for making them **really** hard to use.

I wonder if we should should make this first version of the Path module to work with the PathWithType, and once we have the in-file `module` syntax, we can add a `module Relative` and `module Absolute` that operate on their specific `Path<a, b>` types. That way you can opt-in to more type safe APIs, something like:
```grain
let myPath = Path.fromString("/foo/bar/baz.csv")

let myPath = match (myPath) {
  AbsoluteFile(p) => p,
  AbsoluteDirectory(p) => Path.Absolute.append(p, "index.js"),
  RelativeFile(p) => Path.Absolute.join(Process.cwd(), p),
  RelativeDirectory(p) => Path.Absolute.append(Path.Absolute.join(Process.cwd(), p), "index.js")
}

let myPath = Path.wrapWithType(myPath)

File.readFile(myPath)
```

A lot of this thought process is driven behind what the API for the filesystem will look like...and I think I'd want it to take `PathWithType` for ease of use while still being typed.

What do you think?

Reply:
Hmm, that's an interesting idea; so you're saying for now to make the paths one type and have them be checked dynamically by the functions, then change the functions to throw exceptions or return `Result`s when used on invalid path types?

Reply:
I thought about this some more and thought of another idea that is similar to the one you suggested: for now we stop exposing the `Path<a, b>` type and rename `PathWithType` to just `Path`, changing the types of all functions to use this `Path` and giving `Result`s whenever the path types are invalid for an operation.

Then once the in-file module system drops there could be one submodule called `Typed` that reimplements all of the module's functions but instead using typed paths. This may also be more ergonomic than `Absolute`, `Relative`, etc as all of the "typed" stuff could be in one submodule rather than many.

An example showcasing what the APIs would look like:
```
Path.isDescendantOf: (base: Path, path: Path) -> Result<Bool, IncompatiblePathType>

Path.Typed.isDescendantOf: (base: Typed.Path<a, Directory>, path: Typed.Path<a, c>) -> Bool
```

What do you think?

Reply:
Yeah, let's do that! Oscar was also asking us to use the name `Path` instead of `PathWithType` but I didn't have a good replacement name for `Path<a, b>` (what are you going to name it for now?)

Reply:
Hmm, maybe `TypedPath<a, b>` or something? After the module system is added then we could just do something like this maybe
```
type TypedPath<a, b> = // ...

module Typed {
  type Path<a, b> = TypedPath<a, b>
}
```

Reply:
Sure, let's go with that and people can chime in if they have a better idea 🙇

Reply:
One more question from an API point of view: how should situations where an operation is done on a path type it shouldn't be applicable to be handled? The way I initially outlined would be to enclose all such possible scenarios in a `Result`, returning `Err(IncompatiblePathType)` in situations where the path type is incompatible. However, alternatively perhaps this approach shouldn't always be used in places where it isn't super necessary?

For example, `relativeTo` is an example where doing this approach would make sense, as simply returning `Path` wouldn't be a good idea in error cases for obvious reasons. However, looking at `isDescendantOf` which returns a boolean under normal circumstances, `false` could be just used in these situations instead of an `Err` value, which still makes sense logically.

Overall using this simplification will make the API easier to use but may hide some details about why an operation failed.  Which approach do you think is better?

Reply:
Also on a tangential note, currently `extension` and `stem` return `None` if the file path doesn't have an extension or stem. Should this behavior be changed to return empty string instead in these situations? The decision of which approach to use would also affect these functions

Reply:
I think that returning a Result is absolutely necessary, even if the "okay" case is `Ok(true)`. This should be a relatively fine API because you can pattern match like:
```grain
match (Path.isDescendantOf(...)) {
  Ok(true) => print("hooray"),
  Ok(false) => print("not descendant"),
  Err(_) => print("whoops")
}
```
The reason for this is that (without Result) `false` would be conflated with too many meanings. The specific details will likely be needed by things like the filesystem API.

I'm not sure about the `extension` and `stem`, as I could see it being fine either way. Empty strings might be nicer in that case.

Reply:
Yeah, empty string inside a result; it would give `Err` if the path is a directory and `Ok("")` in the case of a file path with no stem/extension.

Also I just thought of another approach for `isDescendantOf` that may give even more clarity: what do you think about repurposing this function to instead be called something like `ancestry`:
```
enum AncestryStatus {
  Descendant, // if the first argument is the parent of the second
  Ancestor, // if the first argument descends from the second
  Neither, // if the first argument is neither a descendant or ancestor of the second
  DifferentRoots, // if the arguments are both absolute but are on different roots
  IncompatibleBases, // if one argument is relative and other is absolute
}

export let ancestry = (path1, path2) -> AncestryStatus
```
Alternatively, the success cases can be factored out from the error cases:
```
enum AncestryStatus {
  Descendant, // if the first argument is the parent of the second
  Ancestor, // if the first argument descends from the second
  Neither, // if the first argument is neither a descendant or ancestor of the second
}

enum AncestryError {
  DifferentRoots, // if the arguments are both absolute but are on different roots
  IncompatibleBases, // if one argument is relative and other is absolute
}

export let ancestry = (path1, path2) -> Result<AncestryStatus, AncestryError>
```
This would for example allow you to determine the difference between an ancestor and an unrelated path, which both currently result in `false`; let me know what you think.

Reply:
I like it! I think i prefer the `Result<AncestryStatus, AncestryError>` because then you can use all of our Result utilities with it.